### PR TITLE
Update utils based on core refactor and update package.jsons

### PR DESCRIPTION
**Pull request by heath-freenome**
### Reasons for making this change
Partially fixes https://github.com/rjsf-team/react-jsonschema-form/issues/2693 by updating utils types, adding new functions and fixing some bugs

- Updated the main `package.json` to bump the version
- Updated the utils `package.json` to bump everything but react
- Updated `SchemaUtilsType` to add the `getValidator()` and `doesSchemaUtilsDiffer()` functions
  - Updated `createSchemaUtils()` to implement the new functions and the tests to validate them
- Also updated other types to deal with issues found during core refactor
  - Changed `FormValidation` to `ErrorSchema` as needed
  - Updated the `Registry` type to remove `definitions` as it is never used and added the `xxxTemplate` props
  - Updated many interfaces to make previously required props to be optional
  - Switched to using the `React.ComponentType` which incorporates the `FunctionComponent` and `ClassComponent` both
- Fixed a bug in `getDefaultFormState()` by making the `array` defaults use effectively the same logic as it did in `core` but refactored to a function for type safety
  - Updated the tests to add one that verifies the bug is fixed
- Fixed bug in `getSchemaType()` related to incorrectly defaulting to `string` when no type exists
- Also made `uiSchema` optional in `canExpand()`, `getSubmitButtonOptions()`, `getUiOptions()` and `getDisplayLabel()`
- Added new `processSelectValue()` utility, refactored from core's `SelectWidget`, with full tests
- Also updated the `UIOptionsType` to add optional `title` and `description` props, typed to `string` to eliminate the need to type cast them
### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests. I've run `npm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
  - [ ] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/master/CHANGELOG.md) with a description of the PR
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

File: packages/utils/src/processSelectValue.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    return value === 'true';
+  }
+  if (nums.has(type)) {
+    return asNumber(value);
+  }
+
+  // If type is undefined, but an enum is present, try and infer the type from
+  // the enum values
+  if (Array.isArray(schemaEnum)) {
+    if (schemaEnum.every((x: any) => guessType(x) === 'number')) {
```
Comment (nickgros) [inline packages/utils/src/processSelectValue.ts:33]:
```suggestion
    if (schemaEnum.every((x: any) => nums.has(guessType(x)) {
```

File: packages/utils/src/types.ts [hunk 2]
```diff
@@ -163,7 +165,7 @@ export type FieldTemplateProps<T = any, F = any> = {
 export type TitleFieldProps = {
   id: string;
   title: string;
-  required: boolean;
+  required?: boolean;
 };
 
 export type DescriptionFieldProps = {
```
Comment (heath-freenome) [inline packages/utils/src/types.ts:171]: Make description optional

File: packages/utils/src/types.ts [hunk 3]
```diff
@@ -269,17 +271,22 @@ export type UISchemaSubmitButtonOptions = {
 }
 
 export type UIOptionsType = {
-  [key: string]: boolean | number | string | object | any[] | null;
+  /** We know that for title, it will be a string, if it is provided */
+  title?: string;
+  /** We know that for description, it will be a string, if it is provided */
+  description?: string;
```
Comment (heath-freenome) [inline packages/utils/src/types.ts:277]: ui:options?: object

File: packages/utils/test/processSelectValue.test.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    };
+    expect(processSelectValue(schema, 'false')).toBe(false);
+  });
+  it('integer type is converted', () => {
+    const schema: RJSFSchema = {
+      type: 'integer',
+    };
+    expect(processSelectValue(schema, '1')).toEqual(1);
+  });
+  it('enum of number is converted', () => {
```
Comment (heath-freenome) [inline packages/utils/test/processSelectValue.test.ts:48]: Test integer

File: packages/utils/test/processSelectValue.test.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  });
+  it('returns an array of numbers when the type is array and items represents a number', () => {
+    const schema: RJSFSchema = {
+      type: 'array',
+      items: { type: 'number' }
+    };
+    const value = ['1', '1.2'];
+    expect(processSelectValue(schema, value)).toEqual([1, 1.2]);
+  });
+  it('returns an array of strings when the type is array and items do not represent a number', () => {
```
Comment (heath-freenome) [inline packages/utils/test/processSelectValue.test.ts:15]: test integer

File: packages/utils/src/types.ts (hunk 3 above)
Comment (heath-freenome) [reply packages/utils/src/types.ts:277]: On second thought, let's add it later IF we find we need it

**Comment by heath-freenome** [comment]
@nickgros All feedback has been resolved. Hopefully you'll get a chance to review this soon. Got a few more PRs stacking up

**Review by nickgros — APPROVED** [review]

**Pull request merged**